### PR TITLE
feat(runtime): add `swcli runtime lock` command

### PR DESCRIPTION
**Pull request by tianweidut**
## Description
- issues: https://github.com/star-whale/starwhale/issues/741  https://github.com/star-whale/starwhale/issues/764
- support freeze environment: `shell`, `conda env name`, `conda prefix`, `venv dir` 

## Feature Show
- current shell + venv
image

- current shell + conda
image

- name + conda
image

- prefix + conda
image

- prefix + venv
image

## Modules
- [x] Client

## Checklist
- [x] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: client/starwhale/core/runtime/model.py [hunk 1]
```diff
@@ -678,6 +705,103 @@ def activate(cls, workdir: str, yaml_name: str) -> None:
...
+        if runtime_fpath.exists():
+            mode = load_yaml(runtime_fpath).get("mode", PythonRunEnv.VENV)
+
+        if env == RuntimeEnvURIType.SHELL:
+            _current_py_env = guess_current_py_env()
+            if mode and _current_py_env != mode:
+                raise PythonEnvironmentError(
+                    f"runtime.yaml mode:{mode}, shell python env:{_current_py_env}"
+                )
+            mode = mode or _current_py_env
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/core/runtime/model.py:777]: ok

File: client/starwhale/core/runtime/model.py [hunk 2]
```diff
@@ -678,6 +705,103 @@ def activate(cls, workdir: str, yaml_name: str) -> None:
...
+                mode = _dir_mode
+            elif mode != _dir_mode:
+                raise FieldTypeOrValueError(
+                    f"runtime.yaml mode:{mode}, dir mode:{_dir_mode}"
+                )
+
+            env_prefix = env
+        else:
+            if mode == PythonRunEnv.CONDA:
+                env_name = env
```
Comment (jialeicui) [inline client/starwhale/core/runtime/model.py:797]: consider `A` as the value of `--env` param
if conda env A exists and path A exists ether, should we report a conflict

File: client/starwhale/core/runtime/model.py [hunk 3]
```diff
@@ -678,6 +705,103 @@ def activate(cls, workdir: str, yaml_name: str) -> None:
...
+                raise PythonEnvironmentError(
+                    f"runtime.yaml mode:{mode}, shell python env:{_current_py_env}"
+                )
+            mode = mode or _current_py_env
+            env_prefix = get_base_prefix(mode)
+        elif os.path.isdir(env):
+            if os.path.exists(os.path.join(env, "pyvenv.cfg")):
+                _dir_mode = PythonRunEnv.VENV
+            elif os.path.exists(os.path.join(env, "conda-meta")):
+                _dir_mode = PythonRunEnv.CONDA
```
[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/core/runtime/model.py:783]: For example, if the `env` is a virtualenv dir path,  we don't need the `env_name`.

[the remark this answers is not in the document]
Comment (jialeicui) [reply client/starwhale/core/runtime/model.py:783]: My understanding is that when the returned mode is conda, env_name should be set

[the remark this answers is not in the document]
Comment (tianweidut) [reply client/starwhale/core/runtime/model.py:783]: `conda export` has the `--prefix` argument. `env_name` is optional.

[the remark this answers is not in the document]
Comment (jialeicui) [reply client/starwhale/core/runtime/model.py:783]: get it

**Comment by tianweidut** [comment]
@jialeicui  now swcli provides `--name` and `--prefix` arguments to clarify the name and path. The `_fetch_python_env_meta` was removed.  Please review again, thanks

**Comment by jialeicui** [comment]
LGTM

**Review by jialeicui — APPROVED** [review]

**Pull request merged**